pkg/integrations/prometheus/promql/parser.go

```diff
@@ -0,0 +1,564 @@
+	}
+
+	if err != nil {
+		log.Errorf("parsePromQLExprNode: Error parsing promql query: %v", err)
+	}
+
+	return mQueryReqs, queryArithmetic, exit, err
+}
+
+func handleAggregateExpr(expr *parser.AggregateExpr, mQuery *structs.MetricsQuery) {
+	aggFunc := expr.Op.String()
+
+	// Handle parameters if necessary
+	if aggFunc == "quantile" && expr.Param != nil {
+		if param, ok := expr.Param.(*parser.NumberLiteral); ok {
+			mQuery.Aggregator.FuncConstant = param.Val
+		}
+	}
+
+	switch aggFunc {
+	case "avg":
+		mQuery.Aggregator.AggregatorFunction = segutils.Avg
+	case "count":
+		mQuery.Aggregator.AggregatorFunction = segutils.Count
+	case "sum":
+		mQuery.Aggregator.AggregatorFunction = segutils.Sum
+	case "max":
+		mQuery.Aggregator.AggregatorFunction = segutils.Max
+	case "min":
+		mQuery.Aggregator.AggregatorFunction = segutils.Min
+	case "quantile":
+		mQuery.Aggregator.AggregatorFunction = segutils.Quantile
+	default:
+		log.Infof("handleAggregateExpr: using avg aggregator by default for AggregateExpr (got %v)", aggFunc)
+		mQuery.Aggregator = structs.Aggregation{AggregatorFunction: segutils.Avg}
+	}
+
+	// Handle grouping
+	for _, grp := range expr.Grouping {
+		tagFilter := structs.TagsFilter{
+			TagKey:          grp,
+			RawTagValue:     "*",
+			HashTagValue:    xxhash.Sum64String("*"),
+			TagOperator:     segutils.TagOperator(segutils.Equal),
+			LogicalOperator: segutils.And,
+		}
+		mQuery.TagsFilters = append(mQuery.TagsFilters, &tagFilter)
+	}
+	if len(expr.Grouping) > 0 {
+		mQuery.Groupby = true
+	}
+
+	mQueryAgg := &structs.MetricQueryAgg{
+		AggBlockType:    structs.AggregatorBlock,
+		AggregatorBlock: mQuery.Aggregator.Clone(),
+	}
+
+	// If MQueryAggs is nil, set it to the new MetricQueryAgg
+	if mQuery.MQueryAggs == nil {
+		mQuery.MQueryAggs = mQueryAgg
+	} else {
+		// Otherwise, set the new MetricQueryAgg to the head of the chain
+		// and set the current head as the next of the new MetricQueryAgg
+		mQueryAgg.Next = mQuery.MQueryAggs
+		mQuery.MQueryAggs = mQueryAgg
+	}
+
+	// Reset the Aggregator to handle the next aggregation correctly
+	mQuery.Aggregator = structs.Aggregation{}
+}
+
+func handleCallExpr(call *parser.Call, mQuery *structs.MetricsQuery) error {
+	var err error
+	defaultCase := false
+
+	for _, arg := range call.Args {
+		switch arg := arg.(type) {
+		case *parser.MatrixSelector:
+			err = handleCallExprMatrixSelectorNode(call, mQuery)
+		case *parser.VectorSelector:
+			err = handleCallExprVectorSelectorNode(call, mQuery)
+		case *parser.ParenExpr:
+			err = handleCallExprParenExprNode(call, arg, mQuery)
+		case *parser.SubqueryExpr:
+			err = handleCallExprMatrixSelectorNode(call, mQuery)
+		default:
+			defaultCase = true
+		}
+		if err != nil {
+			return fmt.Errorf("handleCallExpr: cannot parse Call Node: %v", err)
+		}
+
+		// Break if the case is not default. As Call Expr can have multiple arguments
+		// And they will be procesed based on the Node type.
+		if !defaultCase {
+			break
+		}
```

Comment:
Switch these `if`s. As they are, if the first loop iteration is non-default and gives an error and the second loop iteration is default, then the error will get printed in both iterations

Alternatively, you may be able to label this loop and do something like `break ArgsLoop` inside the `default` case—I'm not sure

Reply:
I want this loop to exit as soon as it went to a case that is not default. So we want to return the error that comes from a non-default case.

Reply:
Oh my bad. Your code looks correct; I must have read that as `if defaultCase`